crates/compute/src/cpu/layer.rs

```diff
@@ -50,4 +62,276 @@ impl<F: TowerField> ComputeLayer<F> for CpuLayer<F> {
+			.take(n_vars)
+			.collect::<Vec<_>>();
+
+		let mut buffer = vec![BinaryField128b::ZERO; 1 << n_vars];
+		for x_i in &mut buffer[..4] {
+			*x_i = <BinaryField128b as Field>::random(&mut rng);
+		}
+		let mut buffer_clone = buffer.clone();
+
+		fn compute_single_tensor_expand<F, C: ComputeLayer<F>>(
+			compute: &C,
+			coordinates: &[F],
+			buffer: &mut <C::DevMem as ComputeMemory<F>>::FSliceMut<'_>,
+		) {
+			compute
+				.execute(|exec| {
+					compute.tensor_expand(exec, 2, &coordinates[2..], buffer)?;
+					Ok(vec![])
+				})
+				.unwrap();
+		}
+
+		let compute = <CpuLayer<CanonicalTowerFamily>>::default();
+		compute_single_tensor_expand(&compute, &coordinates, &mut buffer.as_mut());
+
+		tensor_prod_eq_ind(2, &mut buffer_clone, &coordinates[2..]).unwrap();
+		assert_eq!(buffer, buffer_clone);
+	}
+
+	#[test]
+	fn test_exec_single_inner_product() {
+		let n_vars = 8;
+
+		let mut rng = StdRng::seed_from_u64(0);
+
+		let a = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
+			.take(1 << (n_vars - <BinaryField128b as ExtensionField<BinaryField16b>>::LOG_DEGREE))
+			.collect::<Vec<_>>();
+		let b = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
+			.take(1 << n_vars)
+			.collect::<Vec<_>>();
+
+		fn compute_single_inner_product<F, C: ComputeLayer<F>>(
+			compute: &C,
+			a_deg: usize,
+			a: <C::DevMem as ComputeMemory<F>>::FSlice<'_>,
+			b: <C::DevMem as ComputeMemory<F>>::FSlice<'_>,
+		) -> Result<F, Error> {
+			let mut results = compute
+				.execute(|exec| {
+					let ret = compute.inner_product(exec, a_deg, a, b)?;
+					Ok(vec![ret])
+				})
+				.unwrap();
+			Ok(results.remove(0))
+		}
+
+		let compute = <CpuLayer<CanonicalTowerFamily>>::default();
+		let actual =
+			compute_single_inner_product(&compute, BinaryField16b::TOWER_LEVEL, &a, &b).unwrap();
+
+		let expected = std::iter::zip(
+			PackedField::iter_slice(
+				<BinaryField128b as PackedExtension<BinaryField16b>>::cast_bases(&a),
+			),
+			&b,
+		)
+		.map(|(a_i, &b_i)| b_i * a_i)
+		.sum::<BinaryField128b>();
+
+		assert_eq!(actual, expected);
+	}
+
+	#[test]
+	fn test_exec_multiple_multilinear_evaluations() {
+		type CL = CpuLayer<CanonicalTowerFamily>;
+
+		let n_vars = 8;
+
+		let mut rng = StdRng::seed_from_u64(0);
+
+		let mle1 = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
+			.take(1 << (n_vars - <BinaryField128b as ExtensionField<BinaryField16b>>::LOG_DEGREE))
+			.collect::<Vec<_>>();
+		let mle2 = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
+			.take(1 << (n_vars - <BinaryField128b as ExtensionField<BinaryField32b>>::LOG_DEGREE))
+			.collect::<Vec<_>>();
+		let coordinates = repeat_with(|| <BinaryField128b as Field>::random(&mut rng))
+			.take(n_vars)
+			.collect::<Vec<_>>();
+
+		let mut eq_ind_buffer = zeroed_vec(1 << n_vars);
+
+		#[allow(clippy::too_many_arguments)]
+		fn compute_multiple_multilinear_evaluations<F, C: ComputeLayer<F>>(
```

Comment:
Slight preference to move these nested functions out to the `tests` module level.